src/misc/ssh/ssh_decode_sequence_multi.c

```diff
@@ -0,0 +1,116 @@
+/* LibTomCrypt, modular cryptographic library -- Tom St Denis
+ *
+ * LibTomCrypt is a library that provides various cryptographic
+ * algorithms in a highly modular and flexible manner.
+ *
+ * The library is free for all purposes without any express
+ * guarantee it works.
+ */
+#include "tomcrypt_private.h"
+#include <stdarg.h>
+
+/**
+   @file ssh_decode_sequence_multi.c
+   SSH data type representation as per RFC4251, Russ Williams
+*/
+
+#ifdef LTC_SSH
+
+/**
+  Decode a SSH sequence using a VA list
+  @param out    [out] Destination for data
+  @param outlen [in/out] Length of buffer and resulting length of output
+  @remark <...> is of the form <type, data> (int, void*) except for string <type, data, size>
+  @return CRYPT_OK on success
+*/
+int ssh_decode_sequence_multi(const unsigned char *in, unsigned long inlen, ...)
+{
+   int           err;
+   ssh_data_type type;
+   va_list       args;
+
+   LTC_ARGCHK(in    != NULL);
+
+   /* Decode values from buffer */
+   va_start(args, inlen);
+   for (;;) {
+      type = (ssh_data_type)va_arg(args, int);
+
+      if (type == LTC_SSHDATA_EOL) {
```

Comment:
> I was mostly concerned about the need for a bunch of variable definitions at the top of the function which would each only be used by a single case. What about switch-case with braced code blocks?

I don't have a problem with both, but it's true - the braced code blocks worsen readability again

Reply:
I guess there's not going to be any way to make it pretty :(

Switch-case with variable definitions at the top of the loop, then? And I can put `type` into the formal arg list, change the `for` to `while (type != LTC_SSHDATA_EOL)` and put the reading of `type` at the end of the block. Sound good?

Any other changes I should make after that?

Reply:
> I guess there's not going to be any way to make it pretty :(

unfortunately not :/



> Switch-case with variable definitions at the top of the loop, then?

I'd say at the top of the function, not the loop



> And I can put `type` into the formal arg list, change the `for` to `while (type != LTC_SSHDATA_EOL)` and put the reading of `type` at the end of the block. Sound good?

why not change the signature to the proposed `int ssh_decode_sequence_multi(const unsigned char *in, unsigned long inlen, int type, void *data, ...)`? 

Should it really be a valid case to call `ssh_decode_sequence_multi(foo, bar, LTC_SSHDATA_EOL)`?

Reply:
I can't put `type` into the formal parameters because the encoding needs to go through the list twice, and can't be `va_start`ed from the middle of the formal parameters, so it would make the signatures of the two methods different :(
And `data` shouldn't really go in there because it then forces things to be `void*` pointers even when they should be integers :(

Reply:
Both valid points I didn't see!

Again, sorry for the noise!

Reply:
But please, still change back to switch-case :)

Reply:
Done, and I've moved the params to the top of the function.